src/shared/install.c

```diff
@@ -66,6 +69,31 @@ typedef struct {
         OrderedHashmap *have_processed;
 } InstallContext;
 
+typedef struct {
+        const char *pattern;
+        int action;
+} PresetRule;
+
+typedef struct {
+        PresetRule *rules;
+        size_t n_rules;
+} Presets;
+
+static inline void presets_freep(Presets *p) {
+        size_t i;
+
+        if (p) {
+                for (i = 0; i < p->n_rules; i++)
+                        if (p->rules[i].pattern)
+                                free((char *)p->rules[i].pattern);
+
+                free(p->rules);
+                p->n_rules = 0;
+        }
```

Comment:
nitpick: I'd always use an "early-exit" strategy, so that the indentation doesn't get this deep. i.e.

```
…
if (!p)
        return;
…
```

early on, then the whole for loop, is indented one level less.

Also, why the "inline"? The only reason why one would use "inline" is in combination with static in header files afaics.